chia/consensus/blockchain.py

```diff
@@ -1069,61 +1070,62 @@ def seen_compact_proofs(self, vdf_info: VDFInfo, height: uint32) -> bool:
         self._seen_compact_proofs.add(pot_tuple)
         return False
 
-    async def get_block_generator(
-        self, block: BlockInfo, additional_blocks: Optional[Dict[bytes32, FullBlock]] = None
-    ) -> Optional[BlockGenerator]:
-        if additional_blocks is None:
-            additional_blocks = {}
-        ref_list = block.transactions_generator_ref_list
-        if block.transactions_generator is None:
-            assert len(ref_list) == 0
-            return None
-        if len(ref_list) == 0:
-            return BlockGenerator(block.transactions_generator, [])
-
-        result: List[bytes] = []
-        previous_br = await self.get_block_record_from_db(block.prev_header_hash)
-        if previous_br is not None and self.height_to_hash(previous_br.height) == block.prev_header_hash:
-            # We are not in a reorg, no need to look up alternate header hashes
-            # (we can get them from height_to_hash)
-            # in the v2 database, we can look up blocks by height directly
-            # (as long as we're in the main chain)
-            result = await self.block_store.get_generators_at(block.transactions_generator_ref_list)
-        else:
-            # First tries to find the blocks in additional_blocks
-            curr = block
-            additional_height_dict = {}
-            while curr.prev_header_hash in additional_blocks:
-                prev: FullBlock = additional_blocks[curr.prev_header_hash]
-                additional_height_dict[prev.height] = prev
-                curr = prev
-
+    async def lookup_block_generators(self, header_hash: bytes32, generator_refs: Set[uint32]) -> Dict[uint32, bytes]:
+
+        generators: Dict[uint32, bytes] = {}
+
+        # if this is empty, we shouldn't have called this function to begin with
+        assert len(generator_refs)
+
+        # The block heights in the transactions_generator_ref_list don't
+        # necessarily refer to the main chain. The generators may be found in 2
+        # different places. A fork of the chain (but in the database) or in
+        # the main chain.
+
+        #              * <- header_hash
+        #              | :
+        # peak -> *    | : reorg_chain
+        #          \   / :
+        #           \ /  :
+        #            *  <- fork point
+        #         :  |
+        #  main   :  |
+        #  chain  :  |
+        #         :  |
+        #         :  * <- genesis
+
+        # If the block is not part of the main chain, we're on a fork, and we
+        # need to find the fork point
+        peak_block = self.try_block_record(header_hash)
+        if peak_block is None:
+            peak_block = await self.get_block_record_from_db(header_hash)
+        assert peak_block is not None
+        if self.height_to_hash(peak_block.height) != header_hash:
             peak: Optional[BlockRecord] = self.get_peak()
```

Comment:
why do you need to use the peak ?
if the block is not in height_to_hash that means the fork is before that block, so traversing from the peak back to that block seems redundant
you would need to pull it out of DB of its not cached anyway and if the its way in the past using the peak will make us do db calls for more blocks

Reply:
you're saying I could build the height-to-hash for the fork without involving the main chain, just traverse from `peak_block` backwards until I get a header hash that matches the `height-to-hash` mapping?

Reply:
I agree that that would be better. I would like to avoid too many distractions right now though, do you think it's OK to revisit this optimization later, I can create a ticket for it

Reply:
im saying this would give the same result but traversing fewer blocks

```suggestion
        peak_block = await self.get_block_record_from_db(header_hash)
        assert peak_block is not None
        if self.height_to_hash(peak_block.height) != header_hash:
            main_chain_block: Optional[BlockRecord] = self.height_to_block_record(peak_block.height)
            reorg_chain: Dict[uint32, bytes32]
            # Then we look up blocks up to fork point one at a time, backtracking
            reorg_chain, _ = await lookup_fork_chain(
                self,
                (main_chain_block.height, main_chain_block.header_hash),
                (peak_block.height, peak_block.header_hash),
                self.constants,
            )
```

Reply:
this performance issue is not new with my patch. It's not obvious fixing it needs to go into this PR

Reply:
oh, I see that this might be a problem with my patch. I already assume that `self.height_to_hash(peak_block.height)` won't throw. It's definitely possible for the fork we're performing the lookup from to have a higher block height than the main chain. That shouldn't be a fatal error.

Reply:
hm. the existing code seems to do this as well though

Reply:
punting on it: https://chia-network.atlassian.net/browse/CHIA-1298